insonmnia/version/version.go

```diff
@@ -0,0 +1,437 @@
+	}
+
+	m.Updated = updated
+
+	return nil
+}
+
+type VersionManager struct {
+	fs             FS
+	clock          func() time.Time
+	version        semver.Version
+	versionFetcher VersionFetcher
+	expireDuration time.Duration
+}
+
+func NewVersionManager(fs FS, options ...Option) *VersionManager {
+	opts := newOptions()
+	for _, o := range options {
+		o(opts)
+	}
+
+	m := &VersionManager{
+		fs:             fs,
+		clock:          opts.Clock,
+		version:        opts.Version,
+		versionFetcher: opts.VersionFetcher,
+		expireDuration: opts.ExpireDuration,
+	}
+
+	return m
+}
+
+// ValidateCurrentVersion validates that the current application version is
+// the latest, reporting errors otherwise.
+//
+// Is is possible that this method fails to open the version file for some
+// reasons, for example because of permissions. It's the user's responsibility
+// either to handle or to ignore the error returned. It's a good idea is to log
+// such errors as warnings, but do not terminate the program.
+func (m *VersionManager) ValidateCurrentVersion(ctx context.Context) error {
+	latestVersion, err := m.latestVersion(ctx)
+	if err != nil {
+		return err
+	}
+
+	if m.version.Compare(latestVersion) != 0 {
+		return &VersionMismatchError{
+			Version:       m.version,
+			LatestVersion: latestVersion,
+		}
+	}
+
+	return nil
+}
+
+func (m *VersionManager) latestVersion(ctx context.Context) (semver.Version, error) {
+	versionDataValue, err := m.versionData()
+	if err != nil {
+		return semver.Version{}, err
+	}
+
+	if versionDataValue == nil {
+		versionDataValue = &versionData{}
+	}
+
+	if m.isExpired(versionDataValue.Updated) {
+		version, err := m.versionFetcher.Update(ctx)
+		if err != nil {
+			return semver.Version{}, &UpdateError{error: err}
+		}
+
+		versionDataValue = &versionData{
+			Version: version,
+			Updated: m.clock(),
+		}
+
+		file, err := m.fs.Create()
+		if err != nil {
+			return semver.Version{}, &OSError{error: err}
+		}
+		defer file.Close()
+
+		data, err := yaml.Marshal(versionDataValue)
+		if err != nil {
+			return semver.Version{}, &CodecError{error: err}
+		}
+
+		n, err := file.Write(data)
+		if err != nil {
+			return semver.Version{}, &OSError{error: err}
+		}
+		if n < len(data) {
+			return semver.Version{}, &OSError{error: io.ErrShortWrite}
+		}
+	}
+
+	return versionDataValue.Version, nil
+}
+
+func (m *VersionManager) versionData() (*versionData, error) {
+	file, err := m.fs.Open()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+
+		return nil, &OSError{error: err}
+	}
+	defer file.Close()
+
+	data, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, &OSError{error: err}
+	}
+
+	var content versionData
+	if err := yaml.Unmarshal(data, &content); err != nil {
+		return nil, &CodecError{error: err}
+	}
+
+	return &content, nil
+}
+
+func (m *VersionManager) isExpired(time time.Time) bool {
+	return time.Add(m.expireDuration).Before(m.clock())
+}
+
+// how to use it:
```

Comment:
Probably will be better to move usage info into the top of the file?

Reply:
Oh, it's just an ancient artifact. I'll drop it.